### PR TITLE
Leave selection unchanged when toggling line comment

### LINKED ISSUE #36349: Leave selection unchanged when applying ToggleLineComment
When user applies ToggleLineComment to a selection, can the selection be left untouched?

--------------------------------------------
#27198 (comment)

Really nice to see this in Visual Studio 16.1!!

Obviously, it will be better when applied to a single-key shortcut instead of a chord.

Also, the changing of caret position and selection is really annoying. It keeps moving my caret to the end of the line, which is not helpful.

Notes:

For ToggleLineComment applied to a single line with no selection, it would be more useful if it moved the caret down one line. (And keep the caret at the same column position)
This could be a setting in Tools->Options (on by default)
Leave the caret where it started. When applying ToggleLineComment there is no need for the selection to change. Also, there is no need for the caret to move. It can stay where it was. Just need to apply comment markers to the start of each line involved in the selection. (Except for the last line if caret at posn 0).

### PR DESCRIPTION
**Pull request by dibarbet**
Forked from https://github.com/dotnet/roslyn/pull/51411

Resolves #36349

### CODE REVIEW

**Comment by dibarbet** [comment]
thanks @ryzngard !

File: src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs [hunk 1]
```diff
@@ -396,7 +396,7 @@ class C
 {
     void M()
     {
-[|        ////var i = 1;
+        [|////var i = 1;
```
Comment (justcla) [inline src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs:399]: This has changed the selection by including the comment markers inside the selection. This is not the behaviour intended.

File: src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs [hunk 2]
```diff
@@ -336,7 +336,7 @@ class C
 {
     void M()
     {
-[|        //var i = 1;
+        [|//var i = 1;
```
Comment (justcla) [inline src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs:339]: The start marker of the selection should remain as it was - immediately left of "var".

File: src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs [hunk 3]
```diff
@@ -106,7 +106,7 @@ class C
 {
     void M()
     {
-[|        //var i = 1;|]
+        [|//var i = 1;|]
```
Comment (justcla) [inline src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs:109]: This has moved the caret. Suggest leaving the caret unchanged (as per spec).
Caret should be hard left against "var".

File: src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs [hunk 4]
```diff
@@ -235,7 +235,7 @@ class C
 {
     void M()
     {
-[|        //var i = 1; // A comment.|]
+        [|//var i = 1; // A comment.|]
```
Comment (justcla) [inline src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs:238]: This has moved the caret and changed the selection. Please leave the selection unchanged. Caret should remain hard left against "var", which is where it was before applying the action.

File: src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs [hunk 5]
```diff
@@ -310,7 +310,7 @@ class C
 {
     void M()
     {
-[|        //string s = '\\';|]
+        [|//string s = '\\';|]
```
Comment (justcla) [inline src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs:313]: This has changed the selection by adding extra chars at the beginning.
Please leave the selection unchanged by keeping the caret immediately to the left of "string".

File: src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs [hunk 6]
```diff
@@ -423,7 +423,7 @@ class C
 {
     void M()
     {
-[|        //var i = 1; // A comment.
+        [|//var i = 1; // A comment.
```
Comment (justcla) [inline src/EditorFeatures/CSharpTest/CommentSelection/CSharpToggleLineCommentCommandHandlerTests.cs:426]: Selection start should remain the same - right next to "var".

**Review by justcla — COMMENTED** [review]
When applying ToggleLineComment to a selection, the start of the selection should not change. Current tests move the caret two places to the left to include the comment markers. This is not expected behaviour.

**Comment by dibarbet** [comment]
@justcla thanks for the feedback.  I'm not totally sure I agree with not including the comment marker.  When there is a subsection of the line selected I think it makes sense to leave the selection as it was (so you can easily change a variable name or something similar).  However having the selection include the comment marker when the whole line is selected seems to make sense to me.  

I'll try out your suggestion and play around with both options for an hour or two to see which one feels better.

**Comment by justcla** [comment]
Thanks Dave. I agree with you.
I spoke out too soon. We are on the same page. Please disregard my comments on that issue.

**Comment by justcla** [comment]
I still think there is an issue with adding comment markers anywhere but the beginning of the line.

When you apply comment markers one line at a time in an indented section, the comments will be scattered.
When you pre-select the area and apply comment markers, they are lined up - but, somewhat randomly in line with whatever indentation was on the first line. This is begging for inconsistency in how and where the comment marker will be applied depending.

Applying line comment markers at the start of the line provides consistency. And furthermore... why not? The user has asked to comment the entire line. So why start a comment some 'random' number of chars in?

**Pull request merged**